README.md
# rsvp-tracker-api
A ReSTful API for an RSVP tracking service written in Clojure. Persistence is provided by Mongodb.


rsvp-tracker-api/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-04-17
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-04-17
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/rsvp-tracker-api/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/rsvp-tracker-api/compare/0.1.0...0.1.1


rsvp-tracker-api/test/rsvp_tracker_api/core_test.clj
(ns rsvp-tracker-api.core-test
  (:require
    [clojure.test :refer [deftest is testing]]
    [cheshire.core :refer [generate-string parse-string]]
    [ring.mock.request :refer [json-body request]]
    [rsvp-tracker-api.core :refer [handler]]))

(def events-endpoint "/api/v1/events")

(def initial-event {"title" "Test Event 1"
                    "location" "Lafayette, LA"
                    "dateStart" "2017-11-21T00:00:00.000Z"
                    "dateEnd" "2017-11-21T04:00:00.000Z"
                    "description" "Live it up"
                    "organizers" ["fex@bar.com"]
                    "invited" ["bob@foo.com" "sarah@foo.com"]
                    "tentative" []
                    "accepted" []
                    "declined" []})

(def event2 {"title" "Test Event 2"
             "location" "Scott, LA"
             "dateStart" "2017-11-27T23:00:00.000Z"
             "dateEnd" "2017-11-28T02:00:00.000Z"
             "description" "Another round"
             "organizers" ["john@foo.com"]
             "invited" ["adam@foo.com" "sarah@foo.com"]
             "tentative" []
             "accepted" []
             "declined" []})

(def updated-event {"title" "Test Event Uno"
                    "location" "Lafayette, LA"
                    "dateStart" "2017-11-21T00:00:00.000Z"
                    "dateEnd" "2017-11-21T04:00:00.000Z"
                    "description" "Live it up"
                    "organizers" ["fex@bar.com"]
                    "invited" ["bob@foo.com" "sarah@foo.com"]
                    "tentative" []
                    "accepted" []
                    "declined" []})

(def invalid-id 111111111111111111111111)

(def create-error (generate-string {:error "Failed to create event"}))

(def view-empty (generate-string []))

(def err-id-not-found (generate-string {:error "Event with given ID does not exist"}))

(def invalid-endpoint (str events-endpoint (format "/%s" invalid-id)))

(defn get-valid-endpoint
  "Returns the events endpoint with the supplied id"
  [id]
  (str events-endpoint (format "/%s" id)))

(defn get-first-event
  "Get first event created this test session (id is random)"
  []
  (let [response (handler (request :get events-endpoint))]
    (first (parse-string (:body response)))))

(deftest api-base
  (testing "Initial bogus JSON data response from /"
    (is (= (handler (request :get "/"))
           {:status 200
            :headers {"Content-Type" "application/json"}
            :body (generate-string {:foo :bar})}))))

(deftest view-all-events-empty
  (testing "View all events when events is empty"
    (is (= (handler (request :get events-endpoint))
           {:status 200
            :headers {"Content-Type" "application/json"}
            :body view-empty}))))

(deftest create-event-valid-data
  (testing "Create an event at /api/v1/events"
    (is (let [response (handler (-> (request :post events-endpoint)
                                    (json-body initial-event)))]
          (= (:status response)
             201)))))

(deftest create-event-invalid-data
  (testing "Fail to create an event from invalid JSON"
    (is (= (handler (-> (request :post events-endpoint)
                        (json-body "foo")))
           {:status 400
            :headers {"Content-Type" "application/json"}
            :body create-error}))))

(deftest view-one-event-exists
  (testing "View an event by id (exists)"
    (is (let [event (get-first-event)
              id (get-in event ["_id"])
              response (handler (request :get (get-valid-endpoint id)))]
          (= (:status response)
             200)))))

(deftest view-one-event-invalid
  (testing "View and event by id (doesn't exist)"
    (is (let [response (handler (request :get invalid-endpoint))]
          (= (:body response)
             "null")))))

(deftest update-event
  (testing "Update first event"
    (is (let [event (get-first-event)
              id (get-in event ["_id"])
              response (handler (-> (request :put (get-valid-endpoint id))
                                    (json-body updated-event)))]
          (and (= (:status response)
                  200)
               (not= (:body response)
                     "null"))))))

(deftest update-event-invalid
  (testing "Update first event"
    (is (let [response (handler (-> (request :put invalid-endpoint)
                                    (json-body updated-event)))]
          (and (= (:status response)
                  200)
               (= (:body response)
                  "null"))))))


rsvp-tracker-api/project.clj
(defproject rsvp-tracker-api "0.0.1"
  :description "A ReSTful API for an RSVP tracking service written in Clojure. Persistence is provided by Mongodb."
  :license {:name "GNU General Public License v3.0"
            :url "https://www.gnu.org/licenses/gpl-3.0.txt"}
  :dependencies [[org.clojure/clojure "1.9.0"]
                 [cheshire "5.8.0"]
                 [clj-time "0.14.3"]
                 [com.novemberain/monger "3.1.0"]
                 [compojure "1.6.1"]
                 [ring/ring-core "1.6.3"]
                 [ring/ring-defaults "0.3.1"]
                 [ring-cors "0.1.12"]]
  :plugins [[lein-ring "0.12.4"]]
  :ring {:handler rsvp-tracker-api.core/handler
         :port 8080}
  :main ^:skip-aot rsvp-tracker-api.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}
             :dev {:dependencies [[proto-repl "0.3.1"]
                                  [ring/ring-mock "0.3.2"]]}})


rsvp-tracker-api/src/rsvp_tracker_api/util.clj
(ns rsvp-tracker-api.util
  (:gen-class)
  (:require
    [cheshire.core :refer [parse-string]]
    [clj-time.local :as ltime]
    [clojure.java.io :as io]
    [clojure.string :as str]))

(defn body-as-string
  "Returns the body of the quest as a string"
  [request]
  (if-let [body (:body request)]
    (condp instance? body
      java.lang.String body
      (slurp (io/reader body)))))

(defn date-str->date
  "Converts the date strings in data into dates for the db"
  [data]
  (assoc data :dateStart (ltime/to-local-date-time (:dateStart data))
              :dateEnd (ltime/to-local-date-time (:dateEnd data))))

(defn parse-json
  "Parses the JSON from the request data"
  [request]
  (when (#{:post :put} (:request-method request))
    (try
      (if-let [body (body-as-string request)]
        (date-str->date (parse-string body true))
        {:message "No body"})
      (catch Exception e
        (.printStackTrace e)))))

(defn get-res-location
  "Generates the url of the given resource"
  [request id]
  (format "%s://%s:%s%s/%s"
          (name (:scheme request))
          (:server-name request)
          (:server-port request)
          (:uri request)
          (str id)))

(defn gen-links
  "Geenerate HATEOAS links for the given resource"
  [request id]
  (let [self (get-res-location request id)
        events (str/replace self (format "/%s" id) "")]
    (assoc {} :self self :events events)))


rsvp-tracker-api/src/rsvp_tracker_api/event.clj
(ns rsvp-tracker-api.event
  (:gen-class)
  (:require
    [cheshire.core :refer [generate-string]]
    [monger.collection :refer [ensure-index find-map-by-id insert remove-by-id update-by-id]]
    [monger.core :refer [connect get-db]]
    [monger.joda-time]
    [monger.json]
    [rsvp-tracker-api.util :refer [gen-links parse-json]])
  (:import
    [org.bson.types ObjectId]))

(def db-name "rsvp-tracker")

(def collection "events")

(def ttl-secs 86400)

(def error-create (generate-string {:error "Failed to create event"}))

(def error-update (generate-string {:error "Failed to update event"}))

(defn create-event
  "Insert and return the event contained within the given context into the db"
  [request]
  (let [conn (connect)
        db (get-db conn db-name)
        id (ObjectId.)
        preData (parse-json request)
        data (assoc preData :_id id :links (gen-links request id))]
    (if (not= 2 (count (keys data)))
      (let [event (insert db collection data)]
        (if (.wasAcknowledged event)
          (do
            (ensure-index db collection (array-map :dateEnd 1) {:expireAfterSeconds ttl-secs})
            (generate-string data))
          error-create))
      error-create)))

(defn get-event
  "Retrieves a specific event given by id"
  [id]
  (let [conn (connect)
        db (get-db conn db-name)]
    (generate-string (find-map-by-id db collection (ObjectId. id)))))

(defn update-event
  "Updates a specific event given by id"
  [id request]
  (let [conn (connect)
        db (get-db conn db-name)
        data (parse-json request)
        oid (ObjectId. id)
        update (update-by-id db collection oid data)]
    (if (.wasAcknowledged update)
      (get-event id)
      error-update)))

(defn delete-event
  "Deletes a specific event given by id"
  [id]
  (let [conn (connect)
        db (get-db conn db-name)
        oid (ObjectId. id)
        delete (remove-by-id db collection oid)]
    (if (.wasAcknowledged delete)
      (generate-string {:completed true})
      (generate-string {:completed false}))))


rsvp-tracker-api/src/rsvp_tracker_api/core.clj
(ns rsvp-tracker-api.core
  (:gen-class)
  (:require
    [cheshire.core :refer [generate-string]]
    [compojure.core :refer [defroutes DELETE GET POST PUT]]
    [ring.middleware.cors :refer [wrap-cors]]
    [ring.middleware.defaults :refer [wrap-defaults api-defaults]]
    [ring.util.response :refer [content-type response status]]
    [rsvp-tracker-api.event :refer [create-event delete-event get-event update-event]]))

(def base-path "/api/v1")

(defn event-created?
  "Adds a Ring HTTP status depending on the success of creating an event"
  [response]
  (if (not= (:body response) (generate-string {:error "Failed to create event"}))
    (status response 201)
    (status response 400)))

(defn event-deleted?
  "Adds a Ring HTTP status depending on the success of deleting an event"
  [response]
  (if (= (:body response) (generate-string {:completed true}))
    (status response 204)
    (status response 400)))

(defroutes app
  (GET "/" []
    (-> (response (generate-string {:foo "bar"}))
        (content-type "application/json")))
  (POST (str base-path "/events") request
    (-> (response (create-event request))
        (content-type "application/json")
        (event-created?)))
  (GET (str base-path "/events/:id") [id]
    (-> (response (get-event id))
        (content-type "application/json")))
  (PUT (str base-path "/events/:id") [id :as request]
    (-> (response (update-event id request))
        (content-type "application/json")))
  (DELETE (str base-path "/events/:id") [id]
    (-> (response (delete-event id))
        (content-type "application/json")
        (event-deleted?))))

(def handler
  (-> app
    (wrap-cors :access-control-allow-origin [#"http://localhost:3000"]
               :access-control-allow-methods [:get :post :put :delete])
    (wrap-defaults api-defaults)))
